server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/mailet/filter/ActionApplier.java

```diff
@@ -182,4 +201,32 @@ private void sendACopy(MailetContext context, MailAddress originalRecipient, Lis
             LifecycleUtil.dispose(copy);
         }
     }
+
+    private Attribute createAttribute(List<MailAddress> forwards) {
+        return new Attribute(AttributeName.FORWARDED_MAIL_ADDRESSES_ATTRIBUTE_NAME,
+            AttributeValue.of(forwards.stream().map(mailAddress -> AttributeValue.of(mailAddress.asString())).collect(ImmutableList.toImmutableList())));
+    }
+
+    private Attribute createAttribute(Set<MailAddress> forwardedMailAddresses, List<MailAddress> newForwards) {
+        return new Attribute(AttributeName.FORWARDED_MAIL_ADDRESSES_ATTRIBUTE_NAME,
+            AttributeValue.of(Stream.concat(forwardedMailAddresses.stream(), newForwards.stream())
+                .map(mailAddress -> AttributeValue.of(mailAddress.asString()))
+                .collect(ImmutableList.toImmutableList())));
+    }
+
+    private List<MailAddress> getNewForwards(List<MailAddress> forwards, Set<MailAddress> forwardedMailAddresses) {
+        List<MailAddress> newForwards = forwards.stream()
+            .filter(mailAddress -> !forwardedMailAddresses.contains(mailAddress))
+            .collect(ImmutableList.toImmutableList());
+        return newForwards;
+    }
+
+    private Set<MailAddress> getForwardedMailAddresses(Optional<Attribute> attribute) {
+        Collection<AttributeValue> attributeValues = (Collection<AttributeValue>) attribute.get().getValue().getValue();
+        Set<MailAddress> forwardedMailAddresses = attributeValues
+            .stream()
+            .map(Throwing.function(attributeValue -> new MailAddress((String) attributeValue.getValue())))
+            .collect(ImmutableSet.toImmutableSet());
+        return forwardedMailAddresses;
+    }
```

Comment:
Any solution to not duplicate that code?

Reply:
The only possible place I can think of to put common code is the mallet api module. However, you said in the comment above that you do not want to pollute the API. Therefore, I do not have any solution :((